cli/tests/test_revset_output.rs

```diff
@@ -496,3 +484,106 @@ fn test_bad_alias_decl() {
       = Redefinition of function parameter
     "###);
 }
+
+#[test]
+fn test_all_modifier() {
+    let test_env = TestEnvironment::default();
+    test_env.jj_cmd_ok(test_env.env_root(), &["git", "init", "repo"]);
+    let repo_path = test_env.env_root().join("repo");
+
+    // Command that accepts single revision by default
+    let stderr = test_env.jj_cmd_failure(&repo_path, &["new", "all()"]);
+    insta::assert_snapshot!(stderr, @r###"
+    Error: Revset "all()" resolved to more than one revision
+    Hint: The revset "all()" resolved to these revisions:
+      qpvuntsm 230dd059 (empty) (no description set)
+      zzzzzzzz 00000000 (empty) (no description set)
+    Hint: Prefix the expression with 'all:' to allow any number of revisions (i.e. 'all:all()').
+    "###);
+    let stderr = test_env.jj_cmd_failure(&repo_path, &["new", "all:all()"]);
+    insta::assert_snapshot!(stderr, @r###"
+    Error: The Git backend does not support creating merge commits with the root commit as one of the parents.
+    "###);
+
+    // Command that accepts multiple revisions by default
+    let stdout = test_env.jj_cmd_success(&repo_path, &["log", "-rall:all()"]);
+    insta::assert_snapshot!(stdout, @r###"
+    @  qpvuntsm test.user@example.com 2001-02-03 08:05:07 230dd059
+    │  (empty) (no description set)
+    ◉  zzzzzzzz root() 00000000
+    "###);
+
+    // Command that accepts only single revision
+    let (_stdout, stderr) = test_env.jj_cmd_ok(&repo_path, &["branch", "create", "-rall:@", "x"]);
+    insta::assert_snapshot!(stderr, @"");
+    let stderr = test_env.jj_cmd_failure(&repo_path, &["branch", "set", "-rall:all()", "x"]);
+    insta::assert_snapshot!(stderr, @r###"
+    Error: Revset "all:all()" resolved to more than one revision
+    Hint: The revset "all:all()" resolved to these revisions:
+      qpvuntsm 230dd059 x | (empty) (no description set)
+      zzzzzzzz 00000000 (empty) (no description set)
+    "###);
+
+    // Template expression that accepts multiple revisions by default
+    let stdout = test_env.jj_cmd_success(&repo_path, &["log", "-Tself.contained_in('all:all()')"]);
+    insta::assert_snapshot!(stdout, @r###"
+    @  true
+    ◉  true
+    "###);
+
+    // Typo
+    let stderr = test_env.jj_cmd_failure(&repo_path, &["new", "ale:x"]);
+    insta::assert_snapshot!(stderr, @r###"
+    Error: Failed to parse revset: Modifier "ale" doesn't exist
+    Caused by:  --> 1:1
+      |
+    1 | ale:x
+      | ^-^
+      |
+      = Modifier "ale" doesn't exist
+    "###);
+
+    // Modifier shouldn't be allowed in aliases
+    let stderr = test_env.jj_cmd_failure(
+        &repo_path,
+        &["new", "x", "--config-toml=revset-aliases.x='all:@'"],
+    );
+    insta::assert_snapshot!(stderr, @r###"
+    Error: Failed to parse revset: Alias "x" cannot be expanded
```

Comment:
Let me know if you'd prefer to leave this for a future PR, but is it possible to have a better error message here ("aliases do not allow the all: modifier")? I think it might be better to parse the `all:` modifier and reject it after it was already parsed.

Reply:
BTW, very good point that we need to parse top-level revsets differently.

Reply:
Perhaps, alias handling can be refactored in a way that `all:` is allowed only if it is expanded at the top level. The current alias processing is a bit wonky because substitution of parsed tree is mixed with context-specific resolution.

It can be addressed later.

Reply:
Add a TODO? I do feel that this error will eventually confuse the people who see it, but I agree this is not urgent. (Hopefully, it won't be many people).

Reply:
(Non-blocking): We could also do something as silly as checking whether an alias starts with `all:` wherever the error that says "Alias 'x' cannot be expanded" is generated.

Reply:
Added comment to the test case.

Better hint could be generated for `whatever & all:whatever` in general, but let's think about that later. It's unlikely to use `all:` out of the command line, I suppose.